File: HeapMaxandMin.py
class MaxHeap:
    def __init__(self):
        self.heap = []
        
    def parent(self, i): return (i - 1) // 2
    def left_child(self, i): return 2 * i + 1
    def right_child(self, i): return 2 * i + 2
    
    def insert(self, value):
        self.heap.append(value)
        self.heapify_up(len(self.heap) - 1)
        
    def heapify_up(self, i):
        while i > 0 and self.heap[i] > self.heap[self.parent(i)]:
            # Swap
            self.heap[i], self.heap[self.parent(i)] = self.heap[self.parent(i)], self.heap[i]
            i = self.parent(i)
            
    def extract_max(self):
        if not self.heap: return None
        max_val = self.heap[0]
        self.heap[0] = self.heap[-1]
        self.heap.pop()
        if self.heap:
            self.heapify_down(0)
        return max_val
        
    def heapify_down(self, i):
        largest = i
        left = self.left_child(i)
        right = self.right_child(i)
        
        if left < len(self.heap) and self.heap[left] > self.heap[largest]:
            largest = left
        if right < len(self.heap) and self.heap[right] > self.heap[largest]:
            largest = right
            
        if largest != i:
            # Swap
            self.heap[i], self.heap[largest] = self.heap[largest], self.heap[i]
            self.heapify_down(largest)
            
    def get_max(self):
        return self.heap[0] if self.heap else None
        
    def display(self):
        print(self.heap)

class MinHeap:
    def __init__(self):
        self.heap = []
        
    def parent(self, i): return (i - 1) // 2
    def left_child(self, i): return 2 * i + 1
    def right_child(self, i): return 2 * i + 2
    
    def insert(self, value):
        self.heap.append(value)
        self.heapify_up(len(self.heap) - 1)
        
    def heapify_up(self, i):
        while i > 0 and self.heap[i] < self.heap[self.parent(i)]:
            # Swap
            self.heap[i], self.heap[self.parent(i)] = self.heap[self.parent(i)], self.heap[i]
            i = self.parent(i)
            
    def extract_min(self):
        if not self.heap: return None
        min_val = self.heap[0]
        self.heap[0] = self.heap[-1]
        self.heap.pop()
        if self.heap:
            self.heapify_down(0)
        return min_val
        
    def heapify_down(self, i):
        smallest = i
        left = self.left_child(i)
        right = self.right_child(i)
        
        if left < len(self.heap) and self.heap[left] < self.heap[smallest]:
            smallest = left
        if right < len(self.heap) and self.heap[right] < self.heap[smallest]:
            smallest = right
            
        if smallest != i:
            # Swap
            self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
            self.heapify_down(smallest)
            
    def get_min(self):
        return self.heap[0] if self.heap else None
        
    def display(self):
        print(self.heap)

if __name__ == "__main__":
    data = [50, 30, 40, 10, 20, 35, 45]
    
    print("MAX-HEAP")
    print("=" * 50)
    
    max_heap = MaxHeap()
    print(f"Memasukkan data: {data}")
    for num in data:
        max_heap.insert(num)
        print(f"Setelah insert {num}: ", end="")
        max_heap.display()
        
    print(f"\nNilai maksimum (root): {max_heap.get_max()}")
    
    print("\nMelakukan extractMax:")
    for _ in range(3):
        removed = max_heap.extract_max()
        print(f"Extract {removed} -> Heap: ", end="")
        max_heap.display()
        
    print("\n")
    print("MIN-HEAP")
    print("=" * 50)
    
    min_heap = MinHeap()
    print(f"Memasukkan data: {data}")
    for num in data:
        min_heap.insert(num)
        print(f"Setelah insert {num}: ", end="")
        min_heap.display()
        
    print(f"\nNilai minimum (root): {min_heap.get_min()}")
    
    print("\nMelakukan extractMin:")
    for _ in range(3):
        removed = min_heap.extract_min()
        print(f"Extract {removed} -> Heap: ", end="")
        min_heap.display()

File: HeapMaxandMin.java
import java.util.ArrayList;
import java.util.Arrays;

class MaxHeap {
    private ArrayList<Integer> heap;
    
    public MaxHeap() {
        heap = new ArrayList<>();
    }
    
    private int parent(int i) { return (i - 1) / 2; }
    private int leftChild(int i) { return 2 * i + 1; }
    private int rightChild(int i) { return 2 * i + 2; }
    
    public void insert(int value) {
        heap.add(value);
        heapifyUp(heap.size() - 1);
    }
    
    private void heapifyUp(int i) {
        while (i > 0 && heap.get(i) > heap.get(parent(i))) {
            int temp = heap.get(i);
            heap.set(i, heap.get(parent(i)));
            heap.set(parent(i), temp);
            i = parent(i);
        }
    }
    
    public Integer extractMax() {
        if (heap.isEmpty()) return null;
        int maxVal = heap.get(0);
        heap.set(0, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);
        if (!heap.isEmpty()) heapifyDown(0);
        return maxVal;
    }
    
    private void heapifyDown(int i) {
        int largest = i;
        int left = leftChild(i);
        int right = rightChild(i);
        if (left < heap.size() && heap.get(left) > heap.get(largest)) largest = left;
        if (right < heap.size() && heap.get(right) > heap.get(largest)) largest = right;
        if (largest != i) {
            int temp = heap.get(i);
            heap.set(i, heap.get(largest));
            heap.set(largest, temp);
            heapifyDown(largest);
        }
    }
    
    public Integer getMax() { return heap.isEmpty() ? null : heap.get(0); }
    public void display() { System.out.println(heap); }
}

class MinHeap {
    private ArrayList<Integer> heap;
    
    public MinHeap() {
        heap = new ArrayList<>();
    }
    
    private int parent(int i) { return (i - 1) / 2; }
    private int leftChild(int i) { return 2 * i + 1; }
    private int rightChild(int i) { return 2 * i + 2; }
    
    public void insert(int value) {
        heap.add(value);
        heapifyUp(heap.size() - 1);
    }
    
    private void heapifyUp(int i) {
        while (i > 0 && heap.get(i) < heap.get(parent(i))) {
            int temp = heap.get(i);
            heap.set(i, heap.get(parent(i)));
            heap.set(parent(i), temp);
            i = parent(i);
        }
    }
    
    public Integer extractMin() {
        if (heap.isEmpty()) return null;
        int minVal = heap.get(0);
        heap.set(0, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);
        if (!heap.isEmpty()) heapifyDown(0);
        return minVal;
    }
    
    private void heapifyDown(int i) {
        int smallest = i;
        int left = leftChild(i);
        int right = rightChild(i);
        if (left < heap.size() && heap.get(left) < heap.get(smallest)) smallest = left;
        if (right < heap.size() && heap.get(right) < heap.get(smallest)) smallest = right;
        if (smallest != i) {
            int temp = heap.get(i);
            heap.set(i, heap.get(smallest));
            heap.set(smallest, temp);
            heapifyDown(smallest);
        }
    }
    
    public Integer getMin() { return heap.isEmpty() ? null : heap.get(0); }
    public void display() { System.out.println(heap); }
}

public class HeapMaxandMin {
    public static void main(String[] args) {
        int[] data = {50, 30, 40, 10, 20, 35, 45};
        
        System.out.println("MAX-HEAP");
        System.out.println("=".repeat(50));
        
        MaxHeap maxHeap = new MaxHeap();
        System.out.println("Memasukkan data: " + Arrays.toString(data));
        for (int num : data) {
            maxHeap.insert(num);
            System.out.print("Setelah insert " + num + ": ");
            maxHeap.display();
        }
        
        System.out.println("\nNilai maksimum (root): " + maxHeap.getMax());
        
        System.out.println("\nMelakukan extractMax:");
        for (int i = 0; i < 3; i++) {
            int removed = maxHeap.extractMax();
            System.out.print("Extract " + removed + " -> Heap: ");
            maxHeap.display();
        }
        
        System.out.println("\n");
        System.out.println("MIN-HEAP");
        System.out.println("=".repeat(50));
        
        MinHeap minHeap = new MinHeap();
        System.out.println("Memasukkan data: " + Arrays.toString(data));
        for (int num : data) {
            minHeap.insert(num);
            System.out.print("Setelah insert " + num + ": ");
            minHeap.display();
        }
        
        System.out.println("\nNilai minimum (root): " + minHeap.getMin());
        
        System.out.println("\nMelakukan extractMin:");
        for (int i = 0; i < 3; i++) {
            int removed = minHeap.extractMin();
            System.out.print("Extract " + removed + " -> Heap: ");
            minHeap.display();
        }
    }
}